File: 55.py
class Solution:
    # @param {integer[]} nums
    # @return {boolean}
    def canJump(self, nums):
        
        nums_len = len(nums)
        if nums_len == 0:
            return True
        cur_idx = 0
        while cur_idx < nums_len-1 :
            jump_len = nums[cur_idx]
            if jump_len == 0:
                return False
            if cur_idx + jump_len >= nums_len-1:
                return True
            
            max_jump = 0
            for idx in range(cur_idx+1 ,cur_idx + jump_len + 1):
                if idx + nums[idx] >= max_jump:
                    max_jump = idx + nums[idx]
                    cur_idx = idx
                    
        return True

File: 55.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int max_idx = 0;
        for(int idx = 0; idx < n-1 ;idx++){
            int step = nums[idx];
            if (step + idx > max_idx)
                max_idx = step + idx;
            if (idx == max_idx)
                return false;
        }
        
        return true;
    }
};